### DISCUSSION TITLE
Mimicking Blue/Green using Canary strategy

### DISCUSSION THREAD

**Discussion by caduri** [argoproj/argo-rollouts#2631] [General]
Hi all, I'm using Blue/Green and I have a scenario where I need to run the verifications in sequence and allow some manual action during the rollout
One of the solutions I thought of was mimicking the Blue/Green strategy using the Canary strategy
I started playing with Argo to see if it can be possible and I have 2 scenarios that I wonder if they are a bug or by design

1. I need to keep the old pods alive till post verifications, using the Canary strategy with `trafficRouting: {}` does it

2. In order to run post verifications I need to change the Stable service to point to the new ReplicaSet by changing `pod-template-hash` field. I found that when the Canary percentage is less than 100% changing `pod-template-hash` is overridden by Argo but when the canary percentage is 100% changing the field is not overridden. 

Are these scenarios bugs?

**Comment by kostis-codefresh** [comment]
> Hi all, I'm using Blue/Green and I have a scenario where I need to run the verifications in sequence and allow some manual action during the rollout

The simplest scenario is stop the new color from auto-promoting, run as many verifications yourself and then promote it. This wouldn't work in your case? Why do you want to use canaries?

**Reply by caduri** [reply]
@kostis-codefresh Thanks for your reply

Running as many verifications won't allow to do it in sequence or allow manual action as I mentioned
Am I missing something?

**Reply by kostis-codefresh** [reply]
I don't understand why you think this is a requirement.  Is there a specific documentation page that you have in mind that says this limitation?

1. You start a blue/green deployment
2. New color is created but is not promoted
3. In your own leisure you can run as many validations as you want in what order you want (using your CI system or something else)
4. You can then promote the rollout via the CLI or the API (manual approval)
5. The new color is promoted

Would that flow work for you? What issues did you find with this?

**Reply by caduri** [reply]
@kostis-codefresh I didn't say its an Argo rollout requirement, sorry if I was misunderstood
I'm doing research on Argo and one of the goals of the research is to check whether this can be done by Argo's blue-green strategy
and if not, can we mimic it using Canary strategy

I know I can achieve it with other tools but this is not the point as I want to check if I can do it with just Argo. 

I found these 2 scenarios and I wonder if they are valid behavior
> 1. I need to keep the old pods alive till post verifications, using the Canary strategy with trafficRouting: {} does it
>
> 2. In order to run post verifications I need to change the Stable service to point to the new ReplicaSet by changing pod-template-hash field. I found that when the Canary percentage is less than 100% changing pod-template-hash is overridden by Argo but when the canary percentage is 100% changing the field is not overridden.

if they are, it will solved my problem
if not and there is no other option then I'll use other tools as you mentioned to achieve it

**Reply by kostis-codefresh** [reply]
You don't need to use other tools for external checks if you don't want to.

You can use https://argoproj.github.io/argo-rollouts/analysis/job/ or https://argoproj.github.io/argo-rollouts/analysis/web/ with plain Argo rollouts. So if you can simply package your checks in a docker image this will work out of the box.with just Argo rollouts and nothing else.

**Reply by caduri** [reply]
@kostis-codefresh I know what kind of verifications I can use with Argo, I wonder if I can run them by sequence and add manual actions during the verifications. 
I understand there is no way to achieve it using blue-green strategy this is why I'm testing Canary strategy, during the research I conducted I found 2 issues that I want to know if they are valid or not and I prefer to focus on them
can you please take a look

> I need to keep the old pods alive till post verifications, using the Canary strategy with trafficRouting: {} does it
>
> In order to run post verifications I need to change the Stable service to point to the new ReplicaSet by changing pod-template-hash field. I found that when the Canary percentage is less than 100% changing pod-template-hash is overridden by Argo but when the canary percentage is 100% changing the field is not overridden.
